lib/sync/storage/dataset-clients.js

```diff
@@ -67,35 +86,47 @@ function doRemoveDatasetClients(datasetClientsToRemove, cb) {
  * @param {Function} cb
  */
 function doRemoveDatasetClientWithRecords(datasetClientId, cb) {
-  debug('doRemoveDatasetClientWithRecords datasetClientId = ' + datasetClientId);
+  debug('doRemoveDatasetClientWithRecords datasetClientId = %s',  datasetClientId);
   async.waterfall([
-    async.apply(doReadDatasetClientWithRecords, datasetClientId),
-    function removeRecords(datasetClientWithRecords, callback) {
-      var records = datasetClientWithRecords.records;
-      if (records.length > 0) {
-        records = syncUtil.convertToObject(records);
-        _.each(records, function(record) {
-          record.op = "delete";
-          delete record._id;
-        });
-        upsertOrDeleteDatasetRecords(datasetClientWithRecords.datasetId, datasetClientId, records, function(err) {
-          return callback(err, datasetClientWithRecords);
-        });
-      } else {
-        return callback(null, datasetClientWithRecords);
-      }
+    async.apply(doReadDatasetClient, datasetClientId),
+    function removeRefs(datasetClientJson, next) {
+      var datasetId = datasetClientJson.datasetId;
+      var recordsCollection = mongoClient.collection(getDatasetRecordsCollectionName(datasetId));
+      recordsCollection.updateMany({'refs': datasetClientId}, {'$pull': {'refs': datasetClientId}}, function(err) {
+        return next(err, datasetClientJson);
+      });
     },
-    function deleteDatasetClient(datasetClientWithRecords, callback) {
+    function deleteDatasetClient(datasetClientJson, next) {
       var col = mongoClient.collection(DATASETCLIENTS_COLLECTION);
-      col.findOneAndDelete({'id': datasetClientId}, function(err, deleted) {
-        return callback(err, deleted);
-      });
+      col.findOneAndDelete({'id': datasetClientId}, next);
     }
   ], function(err, result) {
     if (err) {
       debug('Failed to doRemoveDatasetClientWithRecords due to error %j', err);
```

Comment:
Similar to above comment. Any error here should make it to logs even if the DEBUG flag is not set.